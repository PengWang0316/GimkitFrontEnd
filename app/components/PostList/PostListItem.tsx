import React from 'react';
import {
  List, ListItem, Divider, ListItemText, Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import parse from 'html-react-parser';

import { PostType } from '../../store/Posts/types';

interface Props {
  post: PostType;
  handleClick: (post: PostType) => void;
}

const useStyles = makeStyles({
  rootPaper: {
    width: '100%',
    padding: 15,
  },
  list: {
    width: '100%',
  },
  inline: {
    display: 'inline',
  },
  inlineBold: {
    display: 'inline',
    fontWeight: 'bold',
  },
  itemText: {
    maxHeight: 200,
    overflow: 'hidden',
  },
});
const MAX_CONTENT_LENGTH = 300;
const turncate = (text: string): string => text.length < MAX_CONTENT_LENGTH ? text : `${text.slice(0, MAX_CONTENT_LENGTH)}...`;

export const PostListItem = ({ post, handleClick }: Props) => {
  const classes = useStyles({});

  return (
    <List className={classes.list}>
      <ListItem alignItems="center" button onClick={() => handleClick(post)}>
        <ListItemText
          className={classes.itemText}
          primary={(
            <Typography
              component="span"
              variant="h6"
              className={classes.inlineBold}
              color="textPrimary"
            >
              {`${post.title} - ${new Date(post.date).toLocaleDateString()}`}
            </Typography>
          )}
          secondary={(
            <>
              <Typography
                component="span"
                variant="body2"
                className={classes.inline}
                color="textPrimary"
              >
                <div>{parse(turncate(post.content))}</div>
              </Typography>
            </>
          )}
        />
      </ListItem>
      <Divider component="li" />
    </List>
  );
};

export default PostListItem;
