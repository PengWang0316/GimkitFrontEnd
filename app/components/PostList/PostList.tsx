import React, {
  memo, useEffect, useCallback, useState,
} from 'react';
import { connect } from 'react-redux';
import { Typography, Paper } from '@material-ui/core';
import I18n from '@kevinwang0316/i18n';
import { makeStyles } from '@material-ui/core/styles';
import Pagination from '@kevinwang0316/react-materialui-pagination';

import ReadDialog from './Dialogs/ReadDialog';
import PostListItem from './PostListItem';
import { AppState } from '../../store/ConfigureStore';
import { PostsType, PostType } from '../../store/Posts/types';
import { PostCountType } from '../../store/PostCount/types';
import { fetchPosts as fetchPostsAction } from '../../store/Posts/actions';
import { fetchPostCount as fetchPostCountAction } from '../../store/PostCount/actions';
import { MAX_POSTS_AMOUNT } from '../../config';

interface Props {
  posts: PostsType;
  postCount: PostCountType;
  fetchPosts: (offset: number) => void;
  fetchPostCount: Function;
}

const useStyles = makeStyles({
  rootPaper: {
    width: '100%',
    padding: 15,
  },
});

export const PostList = ({
  posts = null, postCount = null, fetchPosts, fetchPostCount,
}: Props) => {
  const [isOpenRead, setIsOpenRead] = useState(false);
  const [currentPost, setCurrentPost] = useState(null);
  const [offset, setOffset] = useState(0);
  const classes = useStyles({});

  // Initial fetching
  useEffect(() => {
    if (!posts) fetchPosts(0);
    if (postCount === null) fetchPostCount();
  });

  // callback for the reading dialog display
  const handleReadingClose = () => setIsOpenRead(state => !state);

  // callback for click each item
  const itemClickCallback = useCallback((post: PostType) => {
    setCurrentPost(post);
    handleReadingClose();
  }, []);

  // callback for the pagination button click
  const handlePaginationClick = useCallback((newOffset: number) => {
    console.log(newOffset);
    setOffset(newOffset);
  }, []);

  return (
    <div className={classes.rootPaper}>
      <Typography variant="h5" component="h4">
        {I18n.get('postListTitle')}
      </Typography>
      {posts && posts.map(post => (
        <PostListItem
          key={post.id}
          post={post}
          handleClick={itemClickCallback}
        />
      ))}
      {postCount && (
        <Pagination
          offset={offset}
          limit={MAX_POSTS_AMOUNT}
          total={postCount}
          onClick={handlePaginationClick}
        />
      )}
      <ReadDialog
        post={currentPost}
        isOpen={isOpenRead}
        handleClose={handleReadingClose}
      />
    </div>
  );
};

/* istanbul ignore next */
const mapStateToProps = ({ posts, postCount }: AppState) => ({ posts, postCount });
/* istanbul ignore next */
const mapDispatchToProps = { fetchPostCount: fetchPostCountAction, fetchPosts: fetchPostsAction };
export default connect(mapStateToProps, mapDispatchToProps)(memo(PostList));
