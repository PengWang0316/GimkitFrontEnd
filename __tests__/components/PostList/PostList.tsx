import React from 'react';
import renderer from 'react-test-renderer';

import { PostList } from '../../../app/components/PostList/PostList';

jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@kevinwang0316/i18n', () => ({ get: jest.fn().mockReturnValue('I18n text') }));
jest.mock('@kevinwang0316/react-materialui-pagination', () => 'Pagination');
jest.mock('../../../app/components/PostList/Dialogs/ReadDialog', () => 'ReadDialog');
jest.mock('../../../app/components/PostList/Dialogs/DeleteConfirmDialog', () => 'DeleteConfirmDialog');
jest.mock('../../../app/components/PostList/PostListItem', () => 'PostListItem');
jest.mock('../../../app/components/SuccessMessageSnackbar', () => 'SuccessMessageSnackbar');
jest.mock('../../../app/components/SuccessMessageSnackbar', () => 'SuccessMessageSnackbar');
jest.mock('../../../app/components/AddPostDialog', () => 'AddPostDialog');

describe('PostList', () => {
  const defaultProps = {
    posts: [{
      id: 1, content: 'content', title: 'title', date: '01/01/2019',
    }],
    postCount: 1,
    fetchPosts: jest.fn(),
    fetchPostCount: jest.fn(),
  };
  test('snapshot', () => expect(renderer.create(<PostList {...{ ...defaultProps }} />).toJSON()).toMatchSnapshot());
});
