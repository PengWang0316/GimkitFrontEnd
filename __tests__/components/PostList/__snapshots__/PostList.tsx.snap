// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList snapshot 1`] = `
<div
  className="makeStyles-rootPaper-1"
>
  <Typography
    component="h4"
    variant="h5"
  >
    I18n text
  </Typography>
  <PostListItem
    handleClick={[Function]}
    handleDeleteClick={[Function]}
    handleEditCallback={[Function]}
    post={
      Object {
        "content": "content",
        "date": "01/01/2019",
        "id": 1,
        "title": "title",
      }
    }
  />
  <div
    className="makeStyles-paginationDiv-2"
  >
    <Pagination
      limit={5}
      offset={0}
      onClick={[Function]}
      total={1}
    />
  </div>
  <ReadDialog
    handleClose={[Function]}
    isOpen={false}
    post={null}
  />
  <DeleteConfirmDialog
    handleClose={[Function]}
    handleSnackBarCallback={[Function]}
    isOpen={false}
    postId={null}
  />
  <AddPostDialog
    handleClose={[Function]}
    isOpen={false}
    post={null}
  />
  <SuccessMessageSnackbar
    handleClose={[Function]}
    isOpen={false}
    message=""
  />
</div>
`;
